x-pack/platform/plugins/shared/stack_connectors/server/connector_types/mcp/mcp_connector.ts

```diff
@@ -0,0 +1,257 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0; you may not use this file except in compliance with the Elastic License
+ * 2.0.
+ */
+
+import { SubActionConnector, type ServiceParams } from '@kbn/actions-plugin/server';
+import type { AxiosError } from 'axios';
+import type {
+  ConnectorToken,
+  ConnectorTokenClientContract,
+} from '@kbn/actions-plugin/server/types';
+import type {
+  MCPConnectorConfig,
+  MCPConnectorSecrets,
+  CallToolResponse,
+  ListToolsResponse,
+  Tool,
+  CallToolRequest,
+  MCPConnectorHTTPServiceConfig,
+  ContentPart,
+} from '@kbn/mcp-connector-common';
+import { Client } from '@modelcontextprotocol/sdk/client/index.js';
+import { StreamableHTTPClientTransport } from '@modelcontextprotocol/sdk/client/streamableHttp.js';
+import type { Type } from '@kbn/config-schema';
+import { schema } from '@kbn/config-schema';
+import { format } from 'node:util';
+
+export class MCPConnector extends SubActionConnector<MCPConnectorConfig, MCPConnectorSecrets> {
+  private readonly client: Client;
+  private readonly connectorTokenClient: ConnectorTokenClientContract;
+
+  private connected: boolean = false;
+
+  private cachedTools: ListToolsResponse | null = null; // Instance cache
```

Comment:
FMI - Have you tested it, does it actually work that way? I always thought the connector instance was re-created for every sub action call? But given that class registers the sub actions, I realize I was probably wrong.

Reply:
i found this pattern in crowdstrike connector, i didnt test this much yet, but from the logs it looked like its not firing tool discovery requests at every conversation round.